File: elixir/lib/data_structures/deque.ex
defmodule DataStructures.Deque do
  @moduledoc """
  Double-ended queue implementation.
  Time: O(1) amortized for all operations
  Space: O(n)
  """

  defstruct front: [], back: [], count: 0

  @type t :: %__MODULE__{front: list(), back: list(), count: non_neg_integer()}

  @doc "Create an empty deque"
  @spec new() :: t()
  def new, do: %__MODULE__{}

  @doc "Create from list"
  @spec from_list(list()) :: t()
  def from_list(list), do: %__MODULE__{front: list, count: length(list)}

  @doc "Add element to the front"
  @spec push_front(t(), any()) :: t()
  def push_front(%__MODULE__{front: front, back: back, count: count}, value) do
    %__MODULE__{front: [value | front], back: back, count: count + 1}
  end

  @doc "Add element to the back"
  @spec push_back(t(), any()) :: t()
  def push_back(%__MODULE__{front: front, back: back, count: count}, value) do
    %__MODULE__{front: front, back: [value | back], count: count + 1}
  end

  @doc "Remove element from the front"
  @spec pop_front(t()) :: {:ok, any(), t()} | {:error, :empty}
  def pop_front(%__MODULE__{count: 0}), do: {:error, :empty}

  def pop_front(%__MODULE__{front: [], back: back, count: count}) do
    pop_front(%__MODULE__{front: Enum.reverse(back), back: [], count: count})
  end

  def pop_front(%__MODULE__{front: [head | tail], back: back, count: count}) do
    {:ok, head, %__MODULE__{front: tail, back: back, count: count - 1}}
  end

  @doc "Remove element from the back"
  @spec pop_back(t()) :: {:ok, any(), t()} | {:error, :empty}
  def pop_back(%__MODULE__{count: 0}), do: {:error, :empty}

  def pop_back(%__MODULE__{front: front, back: [], count: count}) do
    pop_back(%__MODULE__{front: [], back: Enum.reverse(front), count: count})
  end

  def pop_back(%__MODULE__{front: front, back: [head | tail], count: count}) do
    {:ok, head, %__MODULE__{front: front, back: tail, count: count - 1}}
  end

  @doc "Peek at front element"
  @spec peek_front(t()) :: {:ok, any()} | {:error, :empty}
  def peek_front(%__MODULE__{count: 0}), do: {:error, :empty}
  def peek_front(%__MODULE__{front: [], back: back}), do: {:ok, List.last(back)}
  def peek_front(%__MODULE__{front: [head | _]}), do: {:ok, head}

  @doc "Peek at back element"
  @spec peek_back(t()) :: {:ok, any()} | {:error, :empty}
  def peek_back(%__MODULE__{count: 0}), do: {:error, :empty}
  def peek_back(%__MODULE__{front: front, back: []}), do: {:ok, List.last(front)}
  def peek_back(%__MODULE__{back: [head | _]}), do: {:ok, head}

  @doc "Check if deque is empty"
  @spec empty?(t()) :: boolean()
  def empty?(%__MODULE__{count: 0}), do: true
  def empty?(%__MODULE__{}), do: false

  @doc "Get number of elements"
  @spec size(t()) :: non_neg_integer()
  def size(%__MODULE__{count: count}), do: count

  @doc "Convert to list"
  @spec to_list(t()) :: list()
  def to_list(%__MODULE__{front: front, back: back}) do
    front ++ Enum.reverse(back)
  end
end

File: rust/src/data_structures/deque.rs
use std::collections::VecDeque;

/// A double-ended queue implementation.
/// Time: O(1) for push/pop front/back
/// Space: O(n)
#[derive(Debug, Clone)]
pub struct Deque<T> {
    data: VecDeque<T>,
}

impl<T> Deque<T> {
    pub fn new() -> Self {
        Deque {
            data: VecDeque::new(),
        }
    }

    pub fn with_capacity(capacity: usize) -> Self {
        Deque {
            data: VecDeque::with_capacity(capacity),
        }
    }

    pub fn push_front(&mut self, value: T) {
        self.data.push_front(value);
    }

    pub fn push_back(&mut self, value: T) {
        self.data.push_back(value);
    }

    pub fn pop_front(&mut self) -> Option<T> {
        self.data.pop_front()
    }

    pub fn pop_back(&mut self) -> Option<T> {
        self.data.pop_back()
    }

    pub fn peek_front(&self) -> Option<&T> {
        self.data.front()
    }

    pub fn peek_back(&self) -> Option<&T> {
        self.data.back()
    }

    pub fn peek_front_mut(&mut self) -> Option<&mut T> {
        self.data.front_mut()
    }

    pub fn peek_back_mut(&mut self) -> Option<&mut T> {
        self.data.back_mut()
    }

    pub fn len(&self) -> usize {
        self.data.len()
    }

    pub fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    pub fn clear(&mut self) {
        self.data.clear();
    }

    pub fn get(&self, index: usize) -> Option<&T> {
        self.data.get(index)
    }
}

impl<T> Default for Deque<T> {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_push_pop_front() {
        let mut deque = Deque::new();
        deque.push_front(1);
        deque.push_front(2);

        assert_eq!(deque.pop_front(), Some(2));
        assert_eq!(deque.pop_front(), Some(1));
    }

    #[test]
    fn test_push_pop_back() {
        let mut deque = Deque::new();
        deque.push_back(1);
        deque.push_back(2);

        assert_eq!(deque.pop_back(), Some(2));
        assert_eq!(deque.pop_back(), Some(1));
    }

    #[test]
    fn test_mixed_operations() {
        let mut deque = Deque::new();
        deque.push_back(1);
        deque.push_front(0);
        deque.push_back(2);

        assert_eq!(deque.pop_front(), Some(0));
        assert_eq!(deque.pop_back(), Some(2));
        assert_eq!(deque.pop_front(), Some(1));
    }

    #[test]
    fn test_peek() {
        let mut deque = Deque::new();
        deque.push_back(1);
        deque.push_back(2);

        assert_eq!(deque.peek_front(), Some(&1));
        assert_eq!(deque.peek_back(), Some(&2));
    }

    #[test]
    fn test_get() {
        let mut deque = Deque::new();
        deque.push_back(10);
        deque.push_back(20);
        deque.push_back(30);

        assert_eq!(deque.get(0), Some(&10));
        assert_eq!(deque.get(1), Some(&20));
        assert_eq!(deque.get(2), Some(&30));
        assert_eq!(deque.get(3), None);
    }
}
